Change statefulset update status

dbason opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
This PR changes the status updates so that OnDelete strategy status updates will have the same logic as RollingUpdate.

It updates a comment that has outdated information about the behaviour of scale operations on StatefulSets.

It introduces a new function for calculating the currentRevision of a statefulset so this will be updated where there are no longer any pods with currentRevision when the strategy is OnDelete.

#### Which issue(s) this PR fixes:

Fixes #106055 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
For Statefulsets with an update strategy of OnDelete, the status.currentRevision will be updated when all the pods have been updated to status.updateRevision.  Where there are interim changes made to the StatefulSet the currentRevision will be updated to match the earliest revision in the existing pods.
```

dbason commented:
/check-cla

atiratree commented on pkg/apis/apps/types.go:
```diff
@@ -82,8 +82,8 @@ const (
 	RollingUpdateStatefulSetStrategyType StatefulSetUpdateStrategyType = "RollingUpdate"
-	// OnDeleteStatefulSetStrategyType triggers the legacy behavior. Version
-	// tracking and ordered rolling restarts are disabled. Pods are recreated
+	// OnDeleteStatefulSetStrategyType triggers most of the legacy behavior. Ordered
```
This sentence sounds uncertain and is not great to read without mentioning any specifics.

Since most of the fields were not present in legacy behaviour (https://github.com/kubernetes/kubernetes/commit/1b55f57391f5a648b1f9859118a2d1fef7017ffe#diff-4607db7c58a5ac7a5e4f20c1e2b690c3b3416daafd7b913cda912f16faafef13L121) in my opinion there should not be any hindrance in backwards compatibility except having new updated fields.

atiratree commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -377,13 +377,12 @@ func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) b
-// completeRollingUpdate completes a rolling update when all of set's replica Pods have been updated
+// completeUpdate completes an update when all of set's replica Pods have been updated
 // to the updateRevision. status's currentRevision is set to updateRevision and its' updateRevision
 // is set to the empty string. status's currentReplicas is set to updateReplicas and its updateReplicas
```
the second part of this comment about clearing the values does not seem to hold true

atiratree replied:
please keep this part there: `status's currentReplicas is set to updateReplicas`

atiratree commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -377,13 +377,12 @@ func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) b
-func completeRollingUpdate(set *apps.StatefulSet, status *apps.StatefulSetStatus) {
-	if set.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
-		status.UpdatedReplicas == status.Replicas &&
+func completeUpdate(set *apps.StatefulSet, status *apps.StatefulSetStatus) {
```
It works for the prescribed example, but after testing it manually I have  found a following problem:

```
pods: pod-0(rev1) pod-1(rev2) pod-2(rev3)

StatefulSet:
currentRevision: rev1
updateRevision: rev3
```
`$ kubectl delete pod-0`

```
pods: pod-0(rev3) pod-1(rev2) pod-2(rev3)

StatefulSet:
currentRevision: rev1
updateRevision: rev3
```

even though we have not completed the rev1 is not present anymore, and the current revision should be rev2 now

dbason replied:
As it stands the logic won't touch the currentRevision until everything is updated to the updateRevision.  @atiratree would the expectation be to set currentRevision to rev2 in this scenario?  I believe a RollingUpdate where the spec is updated partway through a rollout would have the same issue, but I can confirm this later today.

atiratree replied:
RollingUpdate tries to update the newest first so there is always a pod present with the oldest revision, so it shouldn't be an issue here

dbason replied:
OK I've introduced a new function for calculating the current version that takes this into consideration.

atiratree replied:
we should get additional opinion on this first, before polishing the code and testing

dbason replied:
I have also added a unit test for the new revision accounting.

atiratree reviewed (commented):
I have reproduced this behaviour and it seems such fix should not hurt anything. But the design doc specifically prohibits this: https://github.com/kubernetes/design-proposals-archive/blob/main/apps/statefulset-update.md#update-completion. I am not sure why. @kow3ns can you please give your opinion on this?

/ok-to-test

dbason commented:
/check-cla

Commit:
Update status for statefulsets with OnDelete strategy

Signed-off-by: Dan Bason <dan.bason@suse.com>

dims commented:
/assign @soltysh

alenkacz commented:
I just bumped into this behavior as well as we're using OnDelete. If there's any way I can help move this PR forward, let me know :)

dbason commented:
> I just bumped into this behavior as well as we're using OnDelete. If there's any way I can help move this PR forward, let me know :)

Just waiting for a review at this point.

dbason commented:
/remove-lifecycle stale

willthames commented:
/remove-lifecycle rotten

dbason commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -555,19 +556,73 @@ func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) b
+// that revision.  If the statefulset has OnDelete update strategy, and there are
+// no longer any pods with the current revision it will get updated if it hasn't
+// completed being rolled out.
+func getCurrentRevision(
```
The testing I added does cover this case.

atiratree commented on pkg/controller/statefulset/stateful_set_control_test.go:
```diff
@@ -727,6 +730,94 @@ func TestStatefulSetControl_getSetRevisions(t *testing.T) {
 		})
+
+	// Test OnDelete revisions
+	testOndelete := func(t *testing.T) {
```
also might be good to name the pods better. eg `pod-1-rev2` or similar

fedebongio commented:
/remove-sig api-machinery

dbason commented:
/remove-sig api-machinery

dbason commented:
@atiratree I've addressed the comments in the latest push.  I've refactored the current revision calculation to something I think should be much clearer.  I've also done the same for the unit test.

I've also updated the OnDelete e2e test to test the new logic.

atiratree commented on pkg/controller/statefulset/stateful_set_control_test.go:
```diff
@@ -734,6 +737,147 @@ func TestStatefulSetControl_getSetRevisions(t *testing.T) {
+		expected:  rev1,
+		set:       set3,
+		pods:      []*v1.Pod{pod0rev1, pod1rev2, pod2rev2},
+		revisions: []*apps.ControllerRevision{rev0, rev1, rev2},
```
can we also add tests for
- list of revisions that is not starting from 0
- list is missing revisions
- pods that are not sorted
- pods only have latest revision
- there is no intersection between pods and revisions

dbason replied:
@atiratree for `list is missing revisions` I assume you're wanting to test where either the CurrentRevision or UpdateRevision is missing from the list of revisions?

dbason replied:
Also for `list of revisions that is not starting from 0` are you aware of a helper method to generate revisions, or should I just be creating statefulsets with arbitrary changes to generate them?

dbason replied:
Likewise with `there is no intersection between pods and revisions` this is a situation that can't occur as the code is currently structured.  There will always be an updateRevision

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -558,11 +558,43 @@ var _ = SIGDescribe("StatefulSet", func() {
 			e2estatefulset.WaitForRunningAndReady(c, 3, ss)
+			ss = waitForStatus(c, ss)
+			currentRevision, updateRevision = ss.Status.CurrentRevision, ss.Status.UpdateRevision
+			framework.ExpectNotEqual(oldestRevision, currentRevision, "Current revision should not be the initial revision")
```
and check the first pod for it

dbason replied:
We do further down.  I will remove this line to avoid confusion

atiratree replied:
@dbason I mean we should check for the exact numbers that we are expecting

dbason commented:
@KnVerey I'll do my best, but unfortunately I've got some other work related stuff on at the moment that I need to take care of.  At this stage I'd say it's 50/50

atiratree commented:
@dbason in case you can't manage, let me know if you would like me to continue on top of your work

fedebongio commented:
/remove-sig api-machinery

dbason commented on pkg/controller/statefulset/stateful_set_control_test.go:
```diff
@@ -734,6 +737,147 @@ func TestStatefulSetControl_getSetRevisions(t *testing.T) {
+	})
+
+	//Test case: pods spread across three revisions
+	set2 := set1.DeepCopy()
```
I had considered this however the problem is that this data is iteratively built on for each test case.  Due to this I took the approach in the existing tests for checking revisions.

atiratree replied:
@dbason  the test data should be split for each testcase - either fully defined in each test object or by having a builder function

please take a look at these tests as an example https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/sync_test.go#L53

dbason commented:
/remove-sig api-machinery

dbason commented:
/remove-sig api-machinery

atiratree commented on pkg/controller/statefulset/stateful_set_control_test.go:
```diff
@@ -734,6 +739,168 @@ func TestStatefulSetControl_getSetRevisions(t *testing.T) {
+			if currentRevision != nil {
+				t.Errorf("for current revision wanted nil got %v", currentRevision)
+			}
+			// Break here to avoid nil pointer errors
```
can we rather use `DeepEqual` here or write the if flow a bit more conscisely?

atiratree commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -555,19 +555,94 @@ func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) b
+			return i
+		}
+	}
+	return -1
```
this was wrong, current should become update rev in that case

dbason commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -555,19 +555,92 @@ func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) b
+
+	// if all of the pods are on the update revision we just return the current revision
+	// and let the complete function update the status
+	if len(revisionsToCheck) == 0 {
```
This is primarily for calculating intermediate states.  If all pods are on the same revision we're not in an intermediate state and the completeUpdate function is responsible for calculating the current revision.

atiratree replied:
I see, can you please better reference `completeUpdate` in a full name in this comment?

atiratree replied:
hm, so the intermediate states are when the pods are missing or are not ready/available? Am I missing some?

dbason replied:
The intermediate states are when pods are spread across 3 revisions and we have to calculate the new CurrentRevision whenever one of the pods is deleted.  Once all the pods are on the UpdateRevision the completeUpdate function will then complete the update and set the CurrentRevision to the UpdateRevision.

atiratree replied:
Okay. I meant mostly under `len(revisionsToCheck) == 0` condition, but nevermind

atiratree commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -555,19 +555,92 @@ func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) b
-	if set.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
-		status.UpdatedReplicas == status.Replicas &&
+func completeUpdate(set *apps.StatefulSet, status *apps.StatefulSetStatus) {
+	if status.UpdatedReplicas == status.Replicas &&
```
@dbason this seems weird. I think this should be `AvailableReplicas` now and it must have been missed when graduating `StatefulSetMinReadySeconds`.

KnVerey commented:
@dbason are you able to continue with this PR?

dbason commented:
/remove-lifecycle rotten

dbason commented:
@KnVerey I haven't had a chance to come back to it sorry (have been caught up with other stuff).  It's 90% there all that is left is cleaning up some of the test code

knelasevero commented:
Hey @dbason , do you mind if I take over this PR? Wanted to tidy last things up and get it merged soon.

knelasevero commented:
Hi, I have opened a different PR with some changes and the rebase from master:

https://github.com/kubernetes/kubernetes/pull/122272

Also splited the simple comment change into a separate PR:

https://github.com/kubernetes/kubernetes/pull/122273
